kconfiglib.py

```diff
@@ -997,34 +998,45 @@ def __repr__(self):
     # File reading
     #
 
-    def _open(self, filename):
+    def _resolve(self, filename, globbing):
         """
-        First tries to open 'filename', then '$srctree/filename' if $srctree
+        First tries with 'filename', then '$srctree/filename' if $srctree
         was set when the configuration was loaded.
         """
+        if os.path.isfile(filename):
+            return [filename]
+
+        if not os.path.isabs(filename) and self.srctree is not None:
+                filename = os.path.join(self.srctree, filename)
+
+        if os.path.isfile(filename):
+            return [filename]
+
+        if globbing:
+            # Try globbing
+            return glob.glob(filename)
```

Comment:
Maybe I haven't understood fully: how could it cause an `IndexError` if I'm calling `self._resolve` with `globbing=False`? No globbing would be attempted in that case and an exception would be raised instead?

Reply:
Ah... sorry, I read it too quickly. Yeah, should be safe in that case.

I was thinking of the `source "nonexistent"` case. I'm guessing the `source` would actually be skipped there now, since `glob("nonexistent")` returns an empty list. It would be better if it still gave an error, to avoid hard-to-diagnose issues.

Reply:
Actually this is intentional. In the Zephyr kernel there are some instances of wildcards that match nothing at all. I was raising an error earlier but had to remove it to be able to process Zephyr's Kconfig tree. I think it's acceptable, given that "no match" is somehow admittable as a result of a wildcard match.

Reply:
It's fine for `source` with wildcards, but when it's just a plain filename, I'd still want it to give an error. Otherwise, `source` statements in other projects that reference nonexistent files would start to fail silently.

Just out of curiosity, do you use a dynamic tree, or is it the same set of files being globbed up every time?

Reply:
> It's fine for source with wildcards, but when it's just a plain filename, I'd still want it to give an error. 

Agreed, the problem is how to find out whether a `source` line contains a wildcard or a plain filename without reimplementing `glob`. I did my best to avoid adding REs or anything else that might be inconsistent with Python's `glob`, and unfortunately there's no "`glob.is_wildcard()`"

Reply:
Yeah, I agree that it gets a bit messy.

Do you think it would be viable to use a different keyword in Zephyr for the globbing `source` (something like `source_glob "foo/*/bar"`)? Then the implementation could be cleanly separated out into a separate code path, and you wouldn't have to worry about getting it to play nice with the existing logic and behavior. Any common parts that still fit could still be factored out.

Reply:
Yes, I think it's a reasonable suggestion. Let me check and get back to you.

Reply:
If you go for a separate keyword, it won't matter much if the behavior is a bit wonky at first either, because it won't have any great effect on how the vanilla Kconfig logic works, and could be improved over time. Gets a bit more iffy when core behavior might change.

Reply:
How about a global flag that defaults to `False` when initializing `Kconfig`? Something like `conf=Kconfig(..., globbing=True)`. This would keep the current behaviour untouched and would simplify our transition to Kconfiglib.

Reply:
This would be transparent for current and future users of Kconfiglib who don't care about globbing, but it would clutter the constructor API a tiny bit when inspecting it. If you don't think this is acceptable then we can also change our Kconfig files.

Reply:
Think I would prefer if the Kconfig files were changed. A flag would need to be supported in the future if people start depending on it, possibly becoming deprecated luggage at some point if it's no longer needed. I try to avoid stuff like that if possible. Plus the globbing, `$srctree`, etc. interactions are pretty tricky to follow already, and would get trickier.